lib/logstash/outputs/s3.rb

```diff
@@ -404,23 +404,36 @@ def configure_upload_workers
         while true do
           @logger.debug("S3: upload worker is waiting for a new file to upload.", :worker_id => worker_id)
 
-          upload_worker
+          begin
+            upload_worker
+          rescue Exception => ex
+            @logger.error('upload_worker unhandled exception', :ex => ex, :backtrace => ex.backtrace)
+            raise LogStash::Error, 'S3: uploader thread exited unexpectedly'
```

Comment:
I don't think this line solve the problem of the thread dying when an exception is raised inside a thread.
I believe we should only catch the exception in the `#upload_worker` method and not in two different places.